libs/SalesforceSDKCommon/SalesforceSDKCommon/Classes/Keychain/KeychainManager.swift

```diff
@@ -0,0 +1,272 @@
+        }
+        
+        if status == errSecItemNotFound {
+            status = SecItemAdd(query as CFDictionary, nil)
+        }
+        
+        //failure to add or update
+        if status != errSecSuccess {
+            return KeychainResult(error: mapError(from: status), status: status)
+        }
+        
+        return self.getValue()
+    }
+    
+    /// Get a value into the keychain for a given identifier.
+    /// - Returns: KeychainResult retrieved for the given identifier
+    func getValue() -> KeychainResult {
+        
+        let additions: [String: Any] = [String(kSecMatchLimit): kSecMatchLimitOne,
+                                        String(kSecReturnAttributes): kCFBooleanTrue as Any,
+                                        String(kSecReturnData): kCFBooleanTrue as Any]
+        
+        var query = secureStoreQueryable.query
+        query.merge(additions) { (current, _) in current }
+        
+        var queryResult: CFTypeRef?
+        let  status = SecItemCopyMatching(query as CFDictionary, &queryResult)
+        
+        guard errSecSuccess == status else {
+            return KeychainResult(error: mapError(from: status), status: status)
+        }
+        
+        guard let item = queryResult as? [String: Any],
+              let resultData = item[String(kSecValueData)] as? Data else {
+            return KeychainResult(data: nil, status: status)
+        }
+        return KeychainResult(data: resultData, status: status)
+        
+    }
+    
+    /// Remove value for a given identifier.
+    /// - Parameter identifier: the key to use.
+    /// - Throws: KeyStoreError wich wraps the underlying error message
+    func removeValue() -> KeychainResult {
+        let query = secureStoreQueryable.query
+        
+        let status = SecItemDelete(query as CFDictionary)
+        guard status == errSecSuccess || status == errSecItemNotFound else {
+            let error = mapError(from: status)
+            return KeychainResult(error: error, status: status)
+        }
+        return KeychainResult(data: nil, status: status)
+    }
+    
+    private func mapError(from status: OSStatus) -> NSError {
+        let message = SecCopyErrorMessageString(status, nil) as String? ?? NSLocalizedString("Unhandled Error", comment: "")
+        let error = NSError(domain: KeychainItemManager.errorDomain, code: Int(status), userInfo: [NSLocalizedDescriptionKey: message, "com.salesforce.security.keychainException.errorCode": status])
+        return error
+    }
+    
+    private func mapError(message: String) -> NSError {
+        let error = NSError(domain: KeychainItemManager.errorDomain, code: -1, userInfo: [NSLocalizedDescriptionKey: message])
+        return error
+    }
+}
+
+
+@objc(SFSDKKeychainResult)
+public class KeychainResult: NSObject {
+    @objc public let success: Bool
+    @objc public var status: OSStatus
+    @objc public var data: Data?
+    @objc public var error: NSError?
```

Comment:
Should `status`, `data`, and `error` be `let` instead? Do they mutate after initialization?

Reply:
When i started these were. They don't need to be. Thanks I will rework this.